components/ai_chat/browser/ai_chat_database_unittest.cc

```diff
@@ -0,0 +1,154 @@
+
+  static std::vector<mojom::ConversationEntryTextPtr> kTexts;
+  kTexts.emplace_back(
+      mojom::ConversationEntryText::New(1, "This is a generated response"));
+  kTexts.emplace_back(
+      mojom::ConversationEntryText::New(2, "This is a re-generated response"));
+
+  InitDBAndDoWork([&]() {
+    db_.AsyncCall(&AIChatDatabase::AddConversation)
+        .WithArgs(mojom::Conversation::New(kConversationId, kCreatedAt,
+                                           "Summarizing Brave",
+                                           GURL("https://brave.com/")))
+        .Then(base::BindLambdaForTesting([&](bool has_run) {
+          db_.AsyncCall(&AIChatDatabase::AddConversationEntry)
+              .WithArgs(kConversationId,
+                        mojom::ConversationEntry::New(
+                            INT64_C(-1), kCreatedAt,
+                            mojom::CharacterType::ASSISTANT, std::move(kTexts)))
+              .Then(base::BindLambdaForTesting([&](bool has_run) {
+                db_.AsyncCall(&AIChatDatabase::GetAllConversations)
+                    .Then(base::BindLambdaForTesting(
+                        [&](std::vector<mojom::ConversationPtr> conversations) {
+                          EXPECT_FALSE(conversations.empty());
+                          EXPECT_EQ(conversations[0]->title,
+                                    "Summarizing Brave");
+                          EXPECT_EQ(conversations[0]->page_url->spec(),
+                                    "https://brave.com/");
+                          EXPECT_EQ(conversations[0]
+                                        ->date.ToDeltaSinceWindowsEpoch()
+                                        .InMicroseconds(),
+                                    kCreatedAt.ToDeltaSinceWindowsEpoch()
+                                        .InMicroseconds());
+
+                          db_.AsyncCall(&AIChatDatabase::GetConversationEntry)
+                              .WithArgs(1)
+                              .Then(base::BindLambdaForTesting(
+                                  [&](std::vector<mojom::ConversationEntryPtr>
+                                          entries) {
+                                    EXPECT_EQ(UINT64_C(1), entries.size());
+                                    EXPECT_EQ(UINT64_C(2),
+                                              entries[0]->texts.size());
+
+                                    EXPECT_EQ(entries[0]->character_type,
+                                              mojom::CharacterType::ASSISTANT);
+                                    EXPECT_EQ(entries[0]->texts[0]->text,
+                                              "This is a generated response");
+                                    EXPECT_EQ(
+                                        entries[0]->texts[1]->text,
+                                        "This is a re-generated response");
+                                  }));
+                        }));
+              }));
+        }));
+  });
+}
```

Comment:
How about a test for how we'll use the database functions:
- Add sample data to the db. It's good to have multiple conversations to properly test the db queries and that they are accurately scoped and limited.
- Get all the conversations and verify the count
- Get the entries for a conversation, verify only the entries for that conversation were received
- Add a new entry, verify getting the entries again has the new entry
- Add a new text to an entry, verify getting the entries again has the new text for that same entry

Reply:
These are all good suggestions and the current state of unittest is doing some of it. 

However, running callbacks one after the other with the current syntax is REALLY painful, to the point where it's hard to read the control flow and manage. If we can figure out how to chain these callbacks or even abstract some of them, these tests can be written trivially.